src/main/kotlin/org/rust/ide/console/RsConsoleRunner.kt

```diff
@@ -0,0 +1,226 @@
+    }
+
+    fun runSync(requestEditorFocus: Boolean) {
+        if (checkNeedInstallEvcxr()) return
+
+        try {
+            initAndRun()
+            ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Connecting to Console", false) {
+                override fun run(indicator: ProgressIndicator) {
+                    indicator.text = "Connecting to console..."
+                    connect()
+                    if (requestEditorFocus) {
+                        consoleView?.requestFocus()
+                    }
+                }
+            })
+        } catch (e: ExecutionException) {
+            LOG.warn("Error running console", e)
+            showErrorsInConsole(e)
+        }
+    }
+
+    fun run(requestEditorFocus: Boolean) {
+        if (checkNeedInstallEvcxr()) return
+        TransactionGuard.submitTransaction(project, Runnable { FileDocumentManager.getInstance().saveAllDocuments() })
+
+        ApplicationManager.getApplication().executeOnPooledThread {
+            ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Connecting to Console", false) {
+                override fun run(indicator: ProgressIndicator) {
+                    indicator.text = "Connecting to console..."
+                    try {
+                        initAndRun()
+                        connect()
+                        if (requestEditorFocus) {
+                            consoleView?.requestFocus()
+                        }
+                    } catch (e: Exception) {
+                        LOG.warn("Error running console", e)
+                        UIUtil.invokeAndWaitIfNeeded(Runnable { showErrorsInConsole(e) })
+                    }
+                }
+            })
+        }
+    }
+
+    private fun checkNeedInstallEvcxr(): Boolean {
+        val needInstallEvcxr = Cargo.checkNeedInstallEvcxr(project)
+        if (needInstallEvcxr) {
+            ApplicationManager.getApplication().invokeLater {
+                RsConsoleToolWindow.getInstance(project).hide()
+            }
+        }
+        return needInstallEvcxr
+    }
+
+    override fun initAndRun() {
+        UIUtil.invokeAndWaitIfNeeded(Runnable {
+            super.initAndRun()
+
+            consoleExecuteActionHandler.sendText(":opt 0\n")
+        })
+    }
+
+    override fun createProcessHandler(process: Process): OSProcessHandler =
+        RsConsoleProcessHandler(
+            process,
+            consoleView,
+            consoleCommunication,
+            commandLine!!.commandLineString,
+            StandardCharsets.UTF_8
+        )
+
+    private fun createCommandLine(): GeneralCommandLine? {
+        val cargoProject = project.cargoProjects.allProjects.firstOrNull()
+        val workingDir = cargoProject?.workingDirectory ?: return null
+        return project.toolchain?.evcxr()?.createCommandLine(workingDir.toFile())
+    }
+
+    override fun createProcess(): Process = commandLine!!.createProcess()
```

Comment:
I'm not sure about ignoring nullability here (and below) too. Is it really OK?

Reply:
Indeed, it can be null if rust toolchain is not defined or evcxr executable not found. So, I modified `createCommandLine` method to throw exception if something goes wrong, and this exception will be catched in `run` or `runSync` methods (and displayed to user)